### PR TITLE
Amplify iOS docs (Getting Started, API, Storage, Analytics)

### PR DESCRIPTION
**Pull request by lawmicha**
https://github.com/undefobj/docs/tree/lawmicha/api-docs/ios

### API (GraphQL)
The docs run through CLI (v4.0.0) to generate a Todo schema. Using `amplify-tools` for model gen, instantiates Amplify and configures API plugin. Guides developer to add unit tests which executes API.mutate, query, query list, and subscribe.
https://github.com/undefobj/docs/blob/lawmicha/api-docs/ios/api.md

TODO: run through different types of schemas.  Post/Comment/Blog schema and many to many
TODO: amplify-tools needs to be updated to provision without sync OR change to run CLI commands. `amplify push` to provision without backend. `amplify codegen models` to get models without sync

### API (Rest)
The doc runs through a very simple example for APIGateway+Lambda+IAM auth. 

### Storage
https://github.com/undefobj/docs/blob/lawmicha/api-docs/ios/storage.md
- TODO: (MInor) Links to AmplifyStorageSample app for more info

### Analytics
https://github.com/undefobj/docs/blob/lawmicha/api-docs/ios/analytics.md

### Getting Started Guide

https://github.com/undefobj/docs/tree/lawmicha/api-docs/ios/start.md
- uses amplify tools, to install CLI, provision API, etc.
- apiPlugin
- shows instantiation and making API calls
- shows configuration file
- link to other resources

Android
https://github.com/undefobj/docs/tree/lawmicha/api-docs/android
https://github.com/undefobj/docs/blob/lawmicha/api-docs/android/analytics.md

### CODE REVIEW

**Comment by lawmicha** [comment]
`npm install -g @aws-amplify/cli` v4.0.0

**Comment by lawmicha** [comment]
App created from storage doc run through https://github.com/lawmicha/docsStorage (is private, just let me know if anyone wants to use, or we can also turn it into a small sample)

**Comment by lawmicha** [comment]
note about API docs - ideally we go straight into using `amplify-tools`

**Comment by lawmicha** [comment]
[masterSchema.graphql.txt](https://github.com/undefobj/docs/files/3906715/masterSchema.graphql.txt)

**Comment by lawmicha** [comment]
some API notes chatting with @mlabieniec 
- Explicit about which schema to use, or better provide a step that adds the schema (in case CLI flow changes in the future) so that they can continue to follow along into the example calls in unit test
- needing to add targets seems like something we should fix in amplify-tools. i added it here because the experience is what it is right now (pinging nikhil L)
- Getting Started guide to focus on DataStore. opportunity to provide a more complex schema example here. 
- we also need to rename the plugin's name fr

**Comment by lawmicha** [comment]
# Datastore with iOS 13

# Getting started

1. Make sure the pod dependencies are pointing to the right place

2. `pod cache clean --all && pod install --repo-update`


# Pod Details


```perl
platform :ios, '13.0'


  pod 'Amplify', :path => '~/aws-amplify/amplify-ios'
  pod 'AWSPluginsCore', :path => '~/aws-amplify/amplify-ios'
  pod 'AmplifyPlugins/AWSDataStoreCategoryPlugin', :path => '~/aws-amplify/amplify-ios'  
  #pod 'amplify-tools'


```

# Amplify CLI details

To instantiate plugin + register models, we need codegen. To codegen, we need schema..

0. Make sure amplify-cli is at the right version, using 4.2.0 right now, update if needed: `npm install @aws-amplify/cli`
1. `amplify add api` and provide schema/ or use default Todo schema

```perl
? Please select from one of the below mentioned services: GraphQL
? Provide API name: testdatastore13
? Choose the default authorization type for the API API key
? Enter a description for the API key: 
? After how many days from now the API key should expire (1-365): 100
? Do you want to configure advanced settings for the GraphQL API No, I am done.
? Do you have an annotated GraphQL schema? No
? Do you want a guided schema creation? Yes
? What best describes your project: Single object with fields (e.g., “Todo” with ID, name, description)
? Do you want to edit the schema now? No
GraphQL schema compiled successfully.

2.  `amplify codegen models`

GraphQL schema compiled successfully.

Edit your schema at /Users/mdlaw/temp/testDataStore13/amplify/backend/api/testdatastore13/schema.graphql or place .graphql files in a directory at /Users/mdlaw/temp/testDataStore13/amplify/backend/api/testdatastore13/schema
Successfully generated models. Generated models can be found /Users/mdlaw/temp/testDataStore13/amplify/generated/models
```

3. open the finder `amplify/generated/models` and drag the model directory into xcode.

4. updated code to 


```swift
import Amplify
import AmplifyPlugins

let dataStorePlugin = AWSDataStoreCategoryPlugin(modelRegistration: AmplifyModels())
        do {
            try Amplify.add(plugin: dataStorePlugin)
            try Amplify.configure()
            print("Amplify initialized")
        } catch {
            print("Failed to configure Amplify \(error)")
        }
```

5. To run successfully, Amplify configure also needs `amplifyconfiguration.json` to exist. If running through these steps from scratch, we don't generate dataStore config from CLI. so i had to update amplifyconfiguration.json manually

From
```
{
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0"
}
```
to
```
{
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0",
    "dataStore": {
        "plugins": {
            "awsDataStoreCategoryPlugin": {
            }
        }
    }
}
```

6. Run the code, success! but with some error as well. Need to investigate
```
2019-11-30 11:10:24.626553-0800 testDataStore13[12734:81789] [AWSDataStoreCategoryPlugin] The operation couldn’t be completed. (Amplify.DataStoreError error 5.)
Success Todo(id: "A730DFA0-4AFC-4D89-A3AA-9857AE973FCC", name: "name", description: Optional("description"))
```

File: ios/analytics.md [hunk 1]
```diff
@@ -1,474 +1,240 @@
...
+                },
+                "autoFlushEventsInterval": 30
+            }
+        }
+    }
 }
 ```
 
-This policy should be applied to roles assigned to the Amazon Cognito identity pool, but you need to replace the `Resource` value with the correct ARN for your Amazon Kinesis or Amazon Kinesis Firehose stream. You can apply policies at the [IAM console](https://console.aws.amazon.com/iam/). To learn more about IAM policies, see [Using IAM](http://docs.aws.amazon.com/IAM/latest/UserGuide/IAM_Introduction.html).
+If you would like to disable automatic flushing of the events to the service, then set this to 0. Then in your app, you can call `Amplify.Analytics.flushEvents()`.
```
Comment (wooj2) [inline ios/analytics.md]: Remove:
Then in your app, you can call...
Put: 
If you do set this to 0, you are responsible for calling `Amplify..flushEvents()` to flush events.

File: ios/analytics.md [hunk 2]
```diff
@@ -1,474 +1,240 @@
...
 
-To learn more about Amazon Kinesis-specific policies, see [Controlling Access to Amazon Kinesis Resources with IAM](http://docs.aws.amazon.com/kinesis/latest/dev/kinesis-using-iam.html).
+### Disable
 
-To learn more about Amazon Kinesis Firehose policies, see [Controlling Access with Amazon Kinesis Firehose](http://docs.aws.amazon.com/firehose/latest/dev/controlling-access.html).
+If some users are in regions which you cannot record events about, simply call `Amplify.Analytics.disable()` and to renable `Amplify.Analytics.enable()`
```
Comment (wooj2) [inline ios/analytics.md]: Not sure if you need to mention "in regions".. i think
To disable analytics, call:
```
Amplify.Analytics.disable()
```
To re-enable, call:
```
Amplify.Analytics.enable()
```

File: ios/analytics.md [hunk 3]
```diff
@@ -1,474 +1,240 @@
...
-
-**Messaging Campaigns** - A campaign sends tailored messages on a schedule that you define. You can create campaigns that send mobile push, email, or SMS messages. To experiment with alternative campaign strategies, set up your campaign as an A/B test, and analyze the results with Amazon Pinpoint analytics.
-
-**Transactional Messages** - Keep your customers informed by sending transactional mobile push and SMS messages—such as new account activation messages, order confirmations, and password reset notifications—to specific users.
-
-**Analyze User Behavior** - You can view trends about your users' level of engagement, purchase activity, and demographics. You can monitor your message traffic with metrics for messages sent and opened. Through the Amazon Pinpoint API, your application can report custom data, which Amazon Pinpoint makes available for analysis.
-
-The Amplify CLI helps setup and configure Pinpoint within your application and connect with the AWS Mobile SDK.
+The Analytics category enables you to collect analytics data for your app. The Analytics category comes with built-in support for [Amazon Pinpoint](#using-amazon-pinpoint)
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply ios/analytics.md]: thanks, updating

File: ios/analytics.md [hunk 4]
```diff
@@ -1,474 +1,240 @@
...
-Use the following steps to add analytics to your mobile app and monitor the results through Amazon Pinpoint.
-
-#### Add Analytics
+Run `amplify console analytics` to open the webpage to view the Pinpoint on AWS Console.
 
-Set up AWS Mobile SDK components as follows.
+### Install Amplify Framework
 
-1. The `Podfile` that you configure to install the AWS Mobile SDK must contain:
+If this is a new project, run `pod init` to create the `Podfile` to use CocoaPods to manage your dependencies. Add the following to the `Podfile`
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply ios/analytics.md]: i figured, if it's an existing project, they would just add it to the Podfile

File: ios/analytics.md [hunk 5]
```diff
@@ -1,474 +1,240 @@
...
-If you encounter an error message that begins `[!] Failed to connect to GitHub to update the CocoaPods/Specs . . .`, and your internet connectivity is working, you may need to [update openssl and Ruby](https://stackoverflow.com/questions/38993527/cocoapods-failed-to-connect-to-github-to-update-the-cocoapods-specs-specs-repo/48962041#48962041).
-
-2. Classes that call Amazon Pinpoint APIs must use the following import statements:
-
-```swift
-/** start code copy **/
-import AWSPinpoint
-import AWSMobileClient
-/** end code copy **/
+Install the dependencies, then close the existing Xcode project if you have it open. We will be using the `xcworkspace` from here on out.
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply ios/analytics.md]: thanks, applying your comments on all docs

File: ios/analytics.md [hunk 6]
```diff
@@ -1,474 +1,240 @@
...
+4. For "Add to targets", make sure the app target is checked off.
+5. Make sure the app builds. You can also use the keyboard shortcut `CMD+B`
 
-    /** start code copy **/
-    var pinpoint: AWSPinpoint?
-    /** end code copy **/
+### Initialize Amplify
 
-    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
+Now we can initialize AWSMobileClient to manage authentication lifecycle, Amplify and the AWSPinpointAnalyticsPlugin.
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply ios/analytics.md]: reducing to 
```
Initialize `AWSMobileClient`, `Amplify`, and `AWSPinpointAnalyticsPlugin`.
```
for now..

File: ios/analytics.md [hunk 7]
```diff
@@ -1,474 +1,240 @@
...
+    ```
 
-### Custom events
+# Use cases
 
-Are nonstandard events that you define by assigning a custom event type. You can add custom attributes and metrics to a custom event.
+The AWSPinpointAnalyticsPlugin provides APIs do the following
 
-On the **Analytics** page in the console, the **Events** tab displays metrics for all custom events that are reported by your app. Use graphs of your custom usage event data in the Amazon Pinpoint console. Visualize how your users' behavior aligns with a model you design using [Amazon Pinpoint Funnel Analytics](https://docs.aws.amazon.com/pinpoint/latest/userguide/analytics-funnels.html), or use [stream the data](https://docs.aws.amazon.com/pinpoint/latest/userguide/analytics-streaming.html) for deeper analysis.
+1. Identify the User to view demographics information about your users
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply ios/analytics.md]: reducing to `1. Identify the user` for now as that is the method signature

File: ios/analytics.md [hunk 8]
```diff
@@ -1,474 +1,240 @@
...
+2. Automatically track app sessions when the user leaves and enters the app. 
+3. Record events to get metrics on usage through the Pinpoint console.
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply ios/analytics.md]: my intent was for the examples to be under "use cases", updated "use cases" to "example use cases" as the title

File: ios/analytics.md [hunk 9]
```diff
@@ -1,474 +1,240 @@
...
+    }
+    wait(for: [completed], timeout: 10)
 }
 ```
 
-### Assigning User IDs to Endpoints
+## App Session Tracking
 
-Assign user IDs to endpoints by doing either of the following:
+The plugin will automatically track when a user opens and closes the app. To view session information, from the terminal run `amplify console analytics` to open the AWS Console for Pinpoint. On the right side, under Analytics, click on Events. Enable filters, you can select `Session Start` and `Session Stop` events to filter on session events.
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply ios/analytics.md]: thanks, updating

File: ios/analytics.md [hunk 10]
```diff
@@ -1,474 +1,240 @@
...
 
-* Manage user sign-up and sign-in with Amazon Cognito user pools.
-* Use the Amazon Pinpoint client to assign user IDs without using Amazon Cognito user pools.
+## Record event
 
-Amazon Cognito user pools are user directories that make it easier to add sign-up and sign-in to your app. When the AWS Mobile SDKs for iOS and Android register an endpoint with Amazon Pinpoint, Amazon Cognito automatically assigns a user ID from the user pool. For more information, see [Using Amazon Pinpoint Analytics with Amazon Cognito User Pools](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-pinpoint-integration.html) in the _Amazon Cognito Developer Guide_.
-
-If you don't want to use Amazon Cognito user pools, you can use the Amazon Pinpoint client in your application to assign user IDs to endpoints.
+You can record events in your app and the plugin will automatically flush the events for you.
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply ios/analytics.md]: yup, we have jazzy docs, can be generated from the amplify-ios directory with `jazzy`

**Comment by lawmicha** [comment]
Currently in discussion around Model gen with/without syncable models. API Category experience may change greatly from using Model based GraphQL APIs to raw documents.

**Comment by lawmicha** [comment]
After how many days from now the API key should expire (1-365): `30`

**Pull request merged**